src/util/sexpr.h

```diff
@@ -10,296 +10,131 @@
+// Forward declarations of templated overloads
+template <typename T>
+void toSExpr(std::ostream& out, const std::vector<T>& v);
 
 /**
- * IOStream manipulator to pretty-print SExprs.
+ * Render an `std::pair` to an s-expression string.
  */
-class CVC4_EXPORT PrettySExprs
+template <typename T1, typename T2>
+void toSExpr(std::ostream& out, const std::pair<T1, T2>& p)
 {
-  /**
-   * The allocated index in ios_base for our setting.
-   */
-  static const int s_iosIndex;
-
-  /**
-   * When this manipulator is used, the setting is stored here.
-   */
-  bool d_prettySExprs;
-
- public:
-  /**
-   * Construct a PrettySExprs with the given setting.
-   */
-  PrettySExprs(bool prettySExprs) : d_prettySExprs(prettySExprs) {}
-
-  inline void applyPrettySExprs(std::ostream& out) {
-    out.iword(s_iosIndex) = d_prettySExprs;
-  }
-
-  static inline bool getPrettySExprs(std::ostream& out) {
-    return out.iword(s_iosIndex);
-  }
-
-  static inline void setPrettySExprs(std::ostream& out, bool prettySExprs) {
-    out.iword(s_iosIndex) = prettySExprs;
-  }
+  out << "(";
+  toSExpr(out, p.first);
+  out << " ";
+  toSExpr(out, p.second);
+  out << ")";
+}
 
-  /**
-   * Set the pretty-sexprs state on the output stream for the current
-   * stack scope.  This makes sure the old state is reset on the
-   * stream after normal OR exceptional exit from the scope, using the
-   * RAII C++ idiom.
-   */
-  class Scope {
-    std::ostream& d_out;
-    bool d_oldPrettySExprs;
-
-   public:
-    inline Scope(std::ostream& out, bool prettySExprs)
-        : d_out(out), d_oldPrettySExprs(PrettySExprs::getPrettySExprs(out)) {
-      PrettySExprs::setPrettySExprs(out, prettySExprs);
+/**
+ * Render an arbitrary iterator range to an s-expression string.
+ */
+template <typename Iterator>
+void toSExpr(std::ostream& out, Iterator begin, Iterator end)
+{
+  out << "(";
+  for (auto it = begin; it != end; ++it)
+  {
+    if (it != begin)
+    {
+      out << " ";
     }
+    toSExpr(out, *it);
+  }
+  out << ")";
+}
 
-    inline ~Scope() { PrettySExprs::setPrettySExprs(d_out, d_oldPrettySExprs); }
-
-  }; /* class PrettySExprs::Scope */
-
-}; /* class PrettySExprs */
+/**
+ * Render a vector to an s-expression string.
+ * Convenience wrapper for `std::vector` around the overload using iterators.
+ */
+template <typename T>
+void toSExpr(std::ostream& out, const std::vector<T>& v)
+{
+  toSExpr(out, v.begin(), v.end());
+}
 
 /**
- * Sets the default pretty-sexprs setting for an ostream.  Use like this:
- *
- *   // let out be an ostream, s an SExpr
- *   out << PrettySExprs(true) << s << endl;
- *
- * The setting stays permanently (until set again) with the stream.
+ * Convert arbitrary data to an s-expression string.
  */
-std::ostream& operator<<(std::ostream& out, PrettySExprs ps);
+template <typename T>
+std::string toSExpr(const T& t)
+{
+  std::stringstream ss;
+  toSExpr(ss, t);
+  return ss.str();
+}
+/**
```

Comment:
Nit: Add a new line.

Reply:
I'm not sure what you mean. Would you add a newline to `ss`? (I'd not do that, as the caller of this function should decide about the context of this s-expression. The s-expression itself has no trailing newline, IMHO).

Reply:
I meant to add a new line between lines 127 and 128 of the code...

Reply:
Ah, yes. Changed it.